course/viewbase.py

```diff
@@ -117,6 +126,15 @@ def get_course_instance_object(self):
             course__url=self.kwargs[self.course_kw],
         )
 
+    def handle_exception(self, exc):
+        if isinstance(exc, TranslationNotFound):
```

Comment:
I tried to test this myself. When I used a non-existing language code in the URL of a course page (`?hl=sv`), the 404 page did not show the available languages. The exception in this function was apparently not of type `TranslationNotFound` and the `languages` template context variable was then not defined. The exception was actually `No CourseInstance matches the given query. <class 'django.http.response.Http404'>` (from a debug print). Maybe that comes from the `LocaleMiddleware` instead of `CourseInstanceBaseMixin`. Can you try this yourself and fix it if the bug is real?

Reply:
I just noticed one important point: `sv` is not enabled in the A+ settings.py. If I use `fi` in an English course, then the 404 page shows the link to the English version of the course page. (`fi` and `en` are the default languages in settings.py.)

Reply:
I think this was a undocumented design decisionI think this was a undocumented design decision.

The `?hl=` is first validated in the middleware, where it has to be something defined in the settings file, hence the `?hl=sv` raises error there. If the variable is valid, then the processing continues to view classes, which include this better solution.

Making the course available to the middleware wouldn't be feasible and ignoring the language error in the middleware might raise other problems.

In short, 404 without language list is application level error, while the other is course level error. Though, that application page _could_ list all configured languages instead of not mentioning anything.

Reply:
I see. Which one is better in the following? If `LocaleMiddleware` raises 404 due to the GET query `hl` parameter (unknown language for the site), then...

* the 404 page shows a link to the same page without any `hl` query parameter. That link would work if the `hl` parameter was the only reason for the 404 error. The user would probably get at least somewhere and understand that the language is not supported. Better to reach some content than to assume that the whole course URL is broken.
* the 404 page shows language links (using `hl` query parameter) to all language versions of the page. All languages would be taken from the site-wide settings.py variable. Not all courses support those languages, so the links would lead to a new 404 page that would finally show working links to the course page.

Reply:
The first alternative without the `hl` seems more reasonable after our daily discussion. Would the link text be something like "open this page in the default language"?

Reply:
One alternative would be doing an automatic redirect, but add like `?msg=automated redirect due to invalid hl parameter`, which would be rendered as a warning flash. I don't remember if we had such a feature yet or not.

Reply:
Since time is running out, we can only either add a link to the 404 page or keep it as is without any link.

Reply:
I personally don't see a huge problem, as it should be less likely to include values which are not valid for any course than have a wrong value for that course (or that specific page in a course).

Hence, I would go what there is now and make issue to fix this better later.

Reply:
I just managed to do the changes regarding the first approach, so I could push those and let me know if it looks good enough or do we want to discard them and go by what we have now.

Reply:
That works too. Btw. do those pages still return 404?

Reply:
> Btw. do those pages still return 404?

They do, so that is good :)